R/step_climate.R

```diff
@@ -338,57 +338,74 @@ print.step_climate <- function(x, width = max(20, options()$width - 30), ...) {
-#' @param .idx the relevant periodic part of time value, e.g. the week number
-#' @param col the list of values indexed by `.idx`
-#' @param weights how much to weigh each particular datapoint
-#' @param aggr the aggregation function, probably Quantile, mean or median
+#' @param idx_in the relevant periodic part of time value, e.g. the week number,
+#'   limited to the relevant range
+#' @param col the list of values indexed by `idx_in`
+#' @param weights how much to weigh each particular datapoint (also indexed by
+#'   `idx_in`)
+#' @param aggr the aggregation function, probably Quantile, mean, or median
 #' @param window_size the number of .idx entries before and after to include in
 #'   the aggregation
-#' @param modulus the maximum value of `.idx`
+#' @param modulus the number of days/weeks/months in the year, not including any
+#'   leap days/weeks
 #' @importFrom lubridate %m-%
 #' @keywords internal
-roll_modular_multivec <- function(col, .idx, weights, aggr, window_size, modulus) {
-  tib <- tibble(col = col, weights = weights, .idx = .idx) |>
+roll_modular_multivec <- function(col, idx_in, weights, aggr, window_size, modulus) {
+  # make a tibble where data gives the list of all datapoints with the
+  # corresponding .idx
+  tib <- tibble(col = col, weights = weights, .idx = idx_in) |>
     arrange(.idx) |>
     tidyr::nest(data = c(col, weights), .by = .idx)
-  out <- double(modulus + 1)
-  for (iter in seq_along(out)) {
-    # +1 from 1-indexing
-    entries <- (iter - window_size):(iter + window_size) %% modulus
-    entries[entries == 0] <- modulus
-    # note that because we are 1-indexing, we're looking for indices that are 1
-    # larger than the actual day/week in the year
-    if (modulus == 365) {
-      # we need to grab just the window around the leap day on the leap day
-      if (iter == 366) {
-        # there's an extra data point in front of the leap day
-        entries <- (59 - window_size):(59 + window_size - 1) %% modulus
-        entries[entries == 0] <- modulus
-        # adding in the leap day itself
-        entries <- c(entries, 999)
-      } else if ((59 %in% entries) || (60 %in% entries)) {
-        # if we're on the Feb/March boundary for daily data, we need to add in the
-        # leap day data
-        entries <- c(entries, 999)
-      }
-    } else if (modulus == 52) {
-      # we need to grab just the window around the leap week on the leap week
-      if (iter == 53) {
-        entries <- (53 - window_size):(53 + window_size - 1) %% 52
-        entries[entries == 0] <- 52
-        entries <- c(entries, 999)
-      } else if ((52 %in% entries) || (1 %in% entries)) {
-        # if we're on the year boundary for weekly data, we need to add in the
-        # leap week data (which is the extra week at the end)
-        entries <- c(entries, 999)
-      }
-    }
-    out[iter] <- with(
+  # storage for the results, includes all possible time indexes
```

Comment:
Note: I'm assuming that tests ensure this works as desired. I'm not checking correctness.